File: app/src/models/Media.ts
import { z } from "zod/v4";
import { Ulid, Timestamp } from "./base";

export const MediaCameraView = z.enum([
  'FRONT',
  'BACK',
  'LEFT',
  'RIGHT',
  'TOP',
  'BOTTOM',
]);

export const MediaMetadata = z.record(z.string(), z.string());

export const MediaStep = z.enum([
  'QUEUING',
  'VIDEO_FETCHING',
  'PRE_PROCESSING',
  'POSE_DETECTING',
  'POST_PROCESSING',
  'FINALIZING',
  'COMPLETED',
]);

const Error = z.object({
  message: z.string(),
  timestamp: Timestamp,
});

export const MediaErrors = z.array(Error);

export const Media = z.object({
  id: Ulid,
  exercise_id: Ulid,
  step: MediaStep,
  camera_view: MediaCameraView,
  original_video_location: z.string(),
  processed_video_location: z.string(),
  pose_detection_model_name: z.string(),
  metadata: MediaMetadata,
  errors: MediaErrors,
  created_at: Timestamp,
  updated_at: Timestamp,
  completed_at: Timestamp,
});

File: backend/src/database/postgres/tables/Media.py
from sqlalchemy import Column, String, DateTime, JSON
from datetime import datetime, UTC
from .base import Base


class MediaTable(Base):
    __tablename__ = "media"

    id = Column(String, primary_key=True)
    exercise_id = Column(String, nullable=False)
    step = Column(String)
    camera_view = Column(String)
    original_video_location = Column(String)
    processed_video_location = Column(String)
    pose_detection_model_name = Column(String)
    media_metadata = Column(name="metadata", type_=JSON)
    errors = Column(JSON)
    angles_of_interest_enum = Column(JSON)
    angles_of_interest = Column(JSON)
    landmark2d_results = Column(JSON)
    landmark3d_results = Column(JSON)
    completed_at = Column(DateTime, nullable=True)
    created_at = Column(
        DateTime,
        default=datetime.now(UTC),
    )
    updated_at = Column(
        DateTime,
        default=datetime.now(UTC),
        onupdate=datetime.now(UTC),
    )
    deleted_at = Column(DateTime, nullable=True)
